README.md
# colatz-length

* コラッツ列の問題解答を（実行時間で）ランキングする。
* dmac/spin で書く web app を学ぶ。
* ryuto の ruby on rails 版と比較する。
* ruby/rails のインストールでヒーコラしているようなので、
  こちらはスタンドアロンのバイナリで行くか？
* 余裕があれば clojure/Duct でも。

works at http://c104.melt.kyutech.ac.jp/ch/ (kyutech only)

---
hiroshi.kimura.0331@gmail.com


ChangeLog
2019-05-20  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.3.8
	LDAP auth.
	table users is dropped.

2019-05-19  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.3.7, 0.3.6
	Bootstrap. table responsive.

	* 0.3.3
	try deploying on vm2019.

	* 0.3.2
	rename project name as colatz-length
	rename remote repository also as
	https://github.com/hkim0331/colatz-length.git

	* 0.3
	start to work on tminit.local. Let's deploy on vm2019.

	* 0.2
	time to move linux.start to work on macos.
		* can upload answers.
		* can launch answers.
		* can retrieve 'time' outputs.
		output format may differ from macos to linux.
		time to move linux.

	* 0.1
	restart developing
	https://github.com/hkim0331/colatz-length.git

2019-05-15  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* 0.0
	start developing


test/length.rkt
#!/usr/bin/env racket
#lang racket
;; tail recursion, length

(define (colatz n)
  (define (c-aux n len)
    (cond
      ((= n 1) len)
      ((even? n) (c-aux (arithmetic-shift n -1) (+ 1 len)))
      (else (c-aux (+ 1 (+ n n n)) (+ 1 len)))))
  (c-aux n 1))

(define (colatz-range n)
  (map (lambda (n) (list (colatz n) n)) (range 1 n)))


(time (argmax first (colatz-range 1000000)))
;;cpu time: 1712 real time: 1713 gc time: 565
;;'(525 837799)

test/memoize.rkt
#!/usr/bin/env racket
#lang racket
(require memoize)

(define (colatz n)
  (define (c-aux n ret)
    (if (= n 1)
     ret
     (c-aux (if (even? n) 
                  (arithmetic-shift n -1)
                  (/ (+ 1 (* 3 n)) 2))
            (if (even? n)
                  (+ 1 ret)
                  (+ 2 ret)))))
  (c-aux n 1))

(define (colatz-range n)
  (map (lambda (n) (list (colatz n) n)) (range (/ n 2) n)))

(time (argmax first (colatz-range 1000000)))


test/tail-recursion.rkt
#!/usr/bin/env racket
#lang racket
;; tail recursion

(define (c-aux n ret)
  (cond
    ((= n 1) ret)
    ((even? n) (c-aux (/ n 2) (cons n ret)))
    (else (c-aux (+ 1 (* n 3)) (cons n ret)))))

(define (colatz n)
  (c-aux n '(1)))

(define (colatz-range n)
  (map (lambda (n) (list (length (colatz n)) n)) (range 1 n)))

(time (argmax first (colatz-range 1000000)))
;;cpu time: 2961 real time: 2971 gc time: 570
;;'(525 837799)



colatz-range.rkt
#!/usr/bin/env racket
#lang racket

(require db racket/system (planet dmac/spin))

(define VERSION "0.3.1")

;;CHANGE
(define DB 
  (first 
    (filter 
      file-exists? 
      '("/Users/hkim/ramdisk/colatz-range-ranks/colatz-range.db"
        "/home/hkim/ramdisk/colatz-range-ranks/colatz-range.db"
        "/opt/colatz-range-ranks/colatz-range.db"))))

(define db (sqlite3-connect #:database DB))

(define (auth? user password)
  (query-maybe-value
    db
    "select id from users where user=$1 and password=$2" user password))

(define (insert user answer msec)
  (query-exec
    db
    "insert into answers (user, answer, msec) values ($1, $2, $3)"
    user answer msec))

(define (answers)
  (query-rows
    db
    "select * from answers where subj='colatz-range' and stat='1' order by msec"))

(define (answer id)
  (query-row
   db
   "select * from answers where id=$1" id))

(define header
  "<!doctype html>
<html>
<head>
<meta charset='utf-8'/>
<meta name='viewport' content='width=device-width, initial-scale=1.0'/>
<link
 rel='stylesheet'
 href='https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css'
 integrity='sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T'
 crossorigin='anonymous'/>
</head>
<body>
<div class='container'>
<h1>Colatz Range</h1>")

(define footer
  "<hr>
programmed by hkimura, 2019-05-15, 2019-05-19.
</div></html>")

(define html
  (lambda body
    (format "~a~a~a"
            header
            (string-join body)
            footer)))

(define (error msg)
  (format
    "<p style='color:red;'>~a</p><p><a href='/upload'>やり直す</a></p>"
    msg))

(define (save s)
  (let ((fname "/tmp/colatz.rkt"))
    (with-output-to-file fname
      (lambda ()
        (displayln s)))
    fname))

(define (msec s)
  (third (string-split s)))

;;FIXME: must return #f if error
(define (try answer)
  (let* ((fname (save answer))
         (s (with-output-to-string
             (lambda ()
               (system (format "racket ~a" fname))))))
        (system (format "rm -f ~a" fname))
        (msec s)))

;; list answers
(define-syntax-rule (inc n)
  (let ()
    (set! n (+ 1 n))
    n))

(get "/"
  (lambda (req)
    (html
     (with-output-to-string
       (lambda ()
         (let ((num 0))
           (displayln "<table>")
           (displayln "<tr><td></td><td>user</td><td>msec</td><td>upload</td>")
           (for ([ans (answers)])
             (displayln
              (format
               "<tr><td>~a</td><td>~a</td><td>~a</td><td><a href='/show/~a'>~a</a></td>"
               (inc num)
               (vector-ref ans 3)
               (vector-ref ans 4)
               (vector-ref ans 0)
               (vector-ref ans 5))))
           (displayln "</table>")
           (displayln "new answer? <a href='/upload'>upload</a>")))))))

(get "/upload"
  (lambda (req)
    (html
      (with-output-to-string
        (lambda ()
          (displayln "<form method='post' action='/submit'>")
          (displayln "<p>name <input name='user'></p>")
          (displayln "<p>password <input name='password' type='password'></p>")
          (displayln "<p>your answer:関数を colayz-range として定義し、<br>")
          (displayln "最後に (time (argmax ...) の計測プログラムを入れてください。</p>")
          (displayln "<textarea name='answer' rows='10' cols='40'>")
          (displayln "(time (argmax first (colatz-range (range 1 1000000))))")
          (displayln "</textarea></p>")
          (displayln "<p><input type='submit'></p>")
          (displayln "</form>"))))))

(post "/submit"
  (lambda (req)
    (html
      (if (not (auth? (params req 'user) (params req 'password)))
        (error "username or password is bad.")
        (let* ((user (params req 'user))
               (answer (params req 'answer))
               (msec (try answer)))
          (if msec
            (begin
            (insert user answer msec)
            (format "<p>thank you. see the <a href='/'>results</a>.</p>"))
            (error "check your answer")))))))

(get "/show/:id"
     (lambda (req)
       (let ((ans (answer (params req 'id))))
         (html
          (with-output-to-string
            (lambda ()
              (display (format "<pre>~a</pre>" (vector-ref ans 6)))))))))

(displayln "server starts at port 8000")
(run #:listen-ip #f #:port 8000)


colatz.rkt
#!/usr/bin/env racket
#lang racket

(require db racket/system net/ldap (planet dmac/spin))

(define VERSION "0.3.8")

;;CHANGE
(define DB
  (first
    (filter
      file-exists?
      `("/srv/colatz-hkim/colatz.db"
        ,(format "~a/colatz.db" (getenv "PWD"))))))

(define db (sqlite3-connect #:database DB))

;; (define (auth? user password)
;;   (query-maybe-value
;;     db
;;     "select id from users where user=$1 and password=$2" user password))

(define (auth? user password)
  (ldap-authenticate "150.69.90.46"
                     389
                     (format "cn=~a,ou=People,dc=local" user)
                     password))

(define (insert user answer msec)
  (query-exec
    db
    "insert into answers (user, answer, msec) values ($1, $2, $3)"
    user answer msec))

(define (answers)
  (query-rows
    db
    "select * from answers where subj='colatz-range' and stat='1' order by msec"))

(define (answer id)
  (query-row
   db
   "select * from answers where id=$1" id))

(define header
  "<!doctype html>
<html>
<head>
<meta charset='utf-8'/>
<meta name='viewport' content='width=device-width, initial-scale=1.0'/>
<link
 rel='stylesheet'
 href='https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css'
 integrity='sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T'
 crossorigin='anonymous'/>
</head>
<body>
<div class='container'>
<h1>Colatz Range</h1>")

(define footer
  "<hr>
programmed by hkimura, 2019-05-15, 2019-05-19.
<a href='https://github.com/hkim0331/colatz-length.git'>github</a>
</div></html>")

(define html
  (lambda body
    (format "~a~a~a"
            header
            (string-join body)
            footer)))

(define (error msg)
  (format
    "<p style='color:red;'>~a</p><p><a href='/upload'>やり直す</a></p>"
    msg))

(define (save s)
  (let ((fname "/tmp/colatz.rkt"))
    (with-output-to-file fname
      (lambda ()
        (displayln s)))
    fname))

(define (msec s)
  (third (string-split s)))

;;FIXME: must return #f if error
(define (try answer)
  (let* ((fname (save answer))
         (s (with-output-to-string
             (lambda ()
               (system (format "racket ~a" fname))))))
        (system (format "rm -f ~a" fname))
        (msec s)))

;; list answers
(define-syntax-rule (inc n)
  (let ()
    (set! n (+ 1 n))
    n))

(get "/"
  (lambda (req)
    (html
     (with-output-to-string
       (lambda ()
         (let ((num 0))
           (displayln "<p>1 ~ 1,000,000 の整数のうちコラッツ列を最長とする n を")
           (displayln "出来るだけ短い時間で求めよ。PE14 と被ったのはたまたまです。</p>")
           (displayln "<div class='table-responsive'>")
           (displayln "<table class='table table-striped'>")
           (displayln "<tr><td></td><td>user</td><td>msec</td><td>upload</td></tr>")
           (for ([ans (answers)])
             (displayln
              (format
               "<tr><td>~a</td><td>~a</td><td stlye='width:6em; text-align:right'>~a</td><td><a href='/show/~a'>~a</a></td></tr>"
               (inc num)
               (vector-ref ans 3)
               (vector-ref ans 4)
               (vector-ref ans 0)
               (vector-ref ans 5))))
           (displayln "</table>")
           (displayln "</div>")
           (displayln "new answer? <a href='/upload' class='btn btn-primary'>upload</a>")))))))

(get "/upload"
  (lambda (req)
    (html
      (with-output-to-string
        (lambda ()
          (displayln "<form method='post' action='/submit'>")
          (displayln "<p>name <input name='user'></p>")
          (displayln "<p>password <input name='password' type='password'></p>")
          (displayln "<p>your answer:関数を colatz-range として定義し、<br>")
          (displayln "最後に (time (argmax ...) の計測プログラムを入れてください。</p>")
          (displayln "<textarea name='answer' rows='10' cols='40'>")
          (displayln "(time (argmax first (colatz-range (range 1 1000000))))")
          (displayln "</textarea></p>")
          (displayln "<p><input type='submit' class='btn btn-primary'></p>")
          (displayln "</form>"))))))

(post "/submit"
  (lambda (req)
    (html
      (if (not (auth? (params req 'user) (params req 'password)))
        (error "username or password is bad.")
        (let* ((user (params req 'user))
               (answer (params req 'answer))
               (msec (try answer)))
          (if msec
            (begin
              (insert user answer msec)
              (format "<p>thank you. see the <a href='/'>results</a>.</p>"))
            (error "check your answer")))))))

(get "/show/:id"
     (lambda (req)
       (let ((ans (answer (params req 'id))))
         (html
          (with-output-to-string
            (lambda ()
              (display (format "<pre><code>~a</code></pre>" (vector-ref ans 6)))))))))

(displayln "server starts at port 8002")
(run #:listen-ip #f #:port 8002)
